File: app/src/main/java/sang/thai/tran/travelcompanion/fragment/RegisterHourlyServiceFragment.java
package sang.thai.tran.travelcompanion.fragment;

import android.content.Context;
import android.os.Bundle;
import androidx.annotation.Nullable;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.TextView;

import butterknife.BindView;
import butterknife.OnClick;
import sang.thai.tran.travelcompanion.R;
import sang.thai.tran.travelcompanion.activity.MainActivity;
import sang.thai.tran.travelcompanion.view.EditTextViewLayout;

import static sang.thai.tran.travelcompanion.utils.AppUtils.openDatePicker;
import static sang.thai.tran.travelcompanion.utils.AppUtils.openTimePicker;

public class RegisterHourlyServiceFragment extends BaseFragment {

    @BindView(R.id.tv_register_service)
    TextView tv_register_service;

    @BindView(R.id.tv_register_service_more)
    TextView tv_register_service_more;

    @BindView(R.id.et_date)
    EditTextViewLayout et_date;

    @BindView(R.id.et_from)
    EditTextViewLayout et_from;

    @BindView(R.id.et_to)
    EditTextViewLayout et_to;

    public static RegisterHourlyServiceFragment newInstance(boolean update) {
        RegisterHourlyServiceFragment infoRegisterFragment = new RegisterHourlyServiceFragment();
        Bundle bundle = new Bundle();
        bundle.putBoolean(MainActivity.NEED_SUPPORT, update);
        infoRegisterFragment.setArguments(bundle);
        return infoRegisterFragment;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        et_date.getEditText().setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                if (hasFocus) {
                    openDepartureDate();
                }
            }
        });

        et_from.getEditText().setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                if (hasFocus) {
                    openFromTime();
                }
            }
        });

        et_to.getEditText().setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                if (hasFocus) {
                    openToTime();
                }
            }
        });

        tv_register_service_more.requestFocus();
        tv_register_service_more.setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                if (hasFocus) {
                    registerServiceMore();
                    tv_register_service_more.clearFocus();
                } else {
                    tv_register_service_more.clearFocus();
                }
            }
        });

        tv_register_service.requestFocus();
        tv_register_service.setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                if (hasFocus) {
                    registerService();
                    tv_register_service.clearFocus();
                } else {
                    tv_register_service.clearFocus();
                }
            }
        });
    }

    @Override
    protected int layoutId() {
        return R.layout.fragment_register_hourly_service;
    }


    @OnClick(R.id.tv_register_service)
    protected void registerService() {
        if (getActivity() == null) {
            return;
        }
        showOptionDialog(tv_register_service
                , getString(R.string.label_for)
                , getActivity().getResources().getTextArray(R.array.register_for_list));

        InputMethodManager imm = (InputMethodManager) getActivity().getSystemService(Context.INPUT_METHOD_SERVICE);
        imm.hideSoftInputFromWindow(tv_register_service.getWindowToken(), 0);
    }

    @OnClick(R.id.tv_register_service_more)
    protected void registerServiceMore() {
        if (getActivity() == null) {
            return;
        }
        showOptionDialog(tv_register_service_more
                , getString(R.string.label_register_service_package_additional)
                , getActivity().getResources().getTextArray(R.array.hourly_service_pkg));

        InputMethodManager imm = (InputMethodManager) getActivity().getSystemService(Context.INPUT_METHOD_SERVICE);
        imm.hideSoftInputFromWindow(tv_register_service_more.getWindowToken(), 0);
    }

    @OnClick(R.id.et_date)
    protected void openDepartureDate() {
        if (getActivity() == null) {
            return;
        }
        openDatePicker(getActivity(), et_date);
    }

    @OnClick(R.id.et_from)
    protected void openFromTime() {
        if (getActivity() == null) {
            return;
        }
        openTimePicker(getActivity(), et_from);
    }

    @OnClick(R.id.et_to)
    protected void openToTime() {
        if (getActivity() == null) {
            return;
        }
        openTimePicker(getActivity(), et_to);
    }

    @OnClick(R.id.email_sign_in_button)
    protected void register() {
        if (getActivity() == null) {
            return;
        }
        ((MainActivity) getActivity()).finishRegister();
    }
}

File: app/src/main/java/sang/thai/tran/travelcompanion/fragment/RegisterHourlyServiceFragment.kt
package sang.thai.tran.travelcompanion.fragment

import android.content.Context
import android.os.Bundle
import android.text.TextUtils
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.TextView
import kotlinx.android.synthetic.main.fragment_register_hourly_service.*
import kotlinx.android.synthetic.main.fragment_register_hourly_service.et_from
import kotlinx.android.synthetic.main.fragment_register_hourly_service.et_to
import kotlinx.android.synthetic.main.fragment_register_hourly_service.ll_parent
import kotlinx.android.synthetic.main.fragment_register_well_companion_page_2.*
import kotlinx.android.synthetic.main.layout_back_next.*
import sang.thai.tran.travelcompanion.R
import sang.thai.tran.travelcompanion.activity.BaseActivity
import sang.thai.tran.travelcompanion.activity.MainActivity
import sang.thai.tran.travelcompanion.interfaces.ResultMultiChoiceDialog
import sang.thai.tran.travelcompanion.model.RegisterModel
import sang.thai.tran.travelcompanion.model.Response
import sang.thai.tran.travelcompanion.retrofit.BaseObserver
import sang.thai.tran.travelcompanion.retrofit.HttpRetrofitClientBase
import sang.thai.tran.travelcompanion.utils.AppConstant
import sang.thai.tran.travelcompanion.utils.AppConstant.*
import sang.thai.tran.travelcompanion.utils.AppUtils.openDatePicker
import sang.thai.tran.travelcompanion.utils.ApplicationSingleton
import sang.thai.tran.travelcompanion.utils.DialogUtils
import sang.thai.tran.travelcompanion.utils.Log
import sang.thai.tran.travelcompanion.view.EditTextViewLayout

class RegisterHourlyServiceFragment : BaseFragment() {

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        et_departure_date?.editText?.onFocusChangeListener = View.OnFocusChangeListener { v, hasFocus ->
            if (hasFocus) {
                openDepartureDate()
            }
        }
        et_departure_date.setOnClickListener { openDepartureDate() }

        et_from?.editText?.onFocusChangeListener = View.OnFocusChangeListener { v, hasFocus ->
            if (hasFocus) {
                openFromTime()
            }
        }
        et_from.setOnClickListener { openFromTime() }

        et_to?.editText?.onFocusChangeListener = View.OnFocusChangeListener { v, hasFocus ->
            if (hasFocus) {
                openToTime()
            }
        }
        et_to.setOnClickListener { openToTime() }

        tv_register_service_more.setOnClickListener { registerServiceMore(API_ADDITIONAL_ASSISTANCE, tv_register_service_more) }

        tv_register_service.setOnClickListener { registerServiceMore(API_SELECTED_ASSISTANCE, tv_register_service) }

        btn_next.setOnClickListener {
            //            openDepartureDate()
            executeRegister()
        }

        btn_back.setOnClickListener {
            //            openDepartureDate()
            createRegisterFlight()
            (activity as BaseActivity).onBackPressed()
        }
        if (tv_register_object != null && ApplicationSingleton.getInstance().data.helperSubjectQualificationList != null)
            setOnClickAndShowDialog(tv_register_object, ApplicationSingleton.getInstance().data.helperSubjectQualificationList!!,
                    object : ResultMultiChoiceDialog {
                        override fun getListSelectedItem(list: List<String>) {
                            ApplicationSingleton.getInstance().professionalRecordsInfoModel?.subject_Qualification_List = list
                        }
                    })
    }

    override fun getApiUrl(): String {
        return API_UPDATE_WELL_TRAINED
    }

    override fun layoutId(): Int {
        return R.layout.fragment_register_hourly_service
    }

    private fun registerServiceMore(url: String, tv: TextView) {
        if (activity == null || isMultiClicked()) {
            return
        }
        val imm = activity?.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        imm.hideSoftInputFromWindow(tv.windowToken, 0)
        Log.d("Sang", "registerServiceMore:")
        HttpRetrofitClientBase.getInstance().executeGet(url,
                ApplicationSingleton.getInstance().token, object : BaseObserver<Response>(true) {
            override fun onSuccess(result: Response, response: String) {
                hideProgressDialog()
                if (activity == null) {
                    return
                }
                if (result.statusCode == AppConstant.SUCCESS_CODE) {
                    Log.d("Sang", "response: $response")
                    lstAssistance = result.result?.data?.list!!

                    result.result?.data?.list?.let { it ->
                        val listString = Array(it.size) { "$it" }
                        for (i in 0 until it.size) {
                            listString[i] = it.get(i).text_VN.toString()
                        }
                        activity?.runOnUiThread {
                            showOptionDialog(tv, tv.text.toString(), listString)
                        }
                    }
                } else {
                    activity?.runOnUiThread { DialogUtils.showAlertDialog(activity, result.message) { dialog, _ -> dialog.dismiss() } }
                }
            }

            override fun onFailure(e: Throwable, errorMsg: String) {
                hideProgressDialog()
                if (!TextUtils.isEmpty(errorMsg)) {
                    activity?.runOnUiThread { DialogUtils.showAlertDialog(activity, errorMsg) { dialog, _ -> dialog.dismiss() } }
                }
            }
        })
    }

    private fun openDepartureDate() {
        if (activity == null || isMultiClicked()) {
            return
        }
        openDatePicker(activity, et_departure_date)
    }

    fun register() {
        if (activity == null) {
            return
        }
        (activity as MainActivity).finishRegister()
    }

    override fun createRegisterFlight(): RegisterModel {
        var registerModel = ApplicationSingleton.getInstance().registerModel
        if (registerModel == null) {
            registerModel = RegisterModel()
        }

        registerModel.id = ApplicationSingleton.getInstance().userInfo.code
        registerModel.contactName = et_full_name?.text

        var wellTrainedObject = tv_register_object?.text.toString()
        wellTrainedObject = wellTrainedObject.substring(wellTrainedObject.length - getString(R.string.label_register_object).length, wellTrainedObject.length)
        registerModel.wellTrainedObject = wellTrainedObject

        var expectAssistance = tv_register_service?.text.toString()
        expectAssistance = expectAssistance.substring(expectAssistance.length - getString(R.string.label_register_service_package).length, expectAssistance.length)
        registerModel.expectAssistance = expectAssistance

        registerModel.departureDateFrom = et_departure_date?.text.toString() + " " + et_departure_time?.text.toString()
        registerModel.pickupPoint = et_pickup_place?.text
        registerModel.visitPlaces = et_visit_place?.text

        var additionalServices = tv_register_service_more?.text.toString()
        additionalServices = additionalServices.substring(additionalServices.length - getString(R.string.label_register_service_package_additional).length, additionalServices.length)
        registerModel.additionalServices = additionalServices

        ApplicationSingleton.getInstance().registerModel = registerModel
        return registerModel
    }


    private fun executeRegister() {
        if (ll_parent == null) {
            return
        }
        val count = ll_parent?.childCount
        for (i in 0 until count!!) {
            val v = ll_parent?.getChildAt(i)
            if (v is EditTextViewLayout) {
                if (TextUtils.isEmpty(v.text)) {
                    showWarningDialog(R.string.label_input_info)
                    return
                }
            }
        }
        registerApi()
    }

    companion object {

        fun newInstance(): RegisterHourlyServiceFragment {
            return RegisterHourlyServiceFragment()
        }
    }
}
